src/Build.UnitTests/Graph/ProjectGraph_Tests.cs

```diff
@@ -842,14 +1094,486 @@ public void TopologicalSortShouldTopologicallySort(Dictionary<int, int[]> edges)
+                _env,
+                1,
+                new[] {2},
+                null,
+                null,
+                innerBuildViaLocalProperty).Path;
+
+            CreateProjectFile(_env, 2);
+
+            var graph = new ProjectGraph(root);
+
+            var dot = graph.ToDot();
+
+            graph.ProjectNodes.Count.ShouldBe(2);
+
+            AssertInnerBuildEvaluation(graph.GraphRoots.First(), false, new Dictionary<string, string>());
+
+            var nonMultitargetingNode = GetFirstNodeWithProjectNumber(graph, 2);
+
+            AssertNonMultitargetingNode(nonMultitargetingNode);
+        }
+
+        [Fact]
+        public void InnerBuildAsRootViaGlobalPropertyShouldNotPropagateInnerBuildPropertyToReference()
+        {
+            var root = CreateProjectFile(_env, 1, new[] {2}, null, null, OuterBuildSpecification).Path;
+            CreateProjectFile(_env, 2);
+
+            var graph = new ProjectGraph(root, new Dictionary<string, string>{{InnerBuildProperty, "foo"}});
+
+            var dot = graph.ToDot();
+
+            graph.ProjectNodes.Count.ShouldBe(2);
+
+            AssertInnerBuildEvaluation(graph.GraphRoots.First(), true, new Dictionary<string, string>());
+
+            var nonMultitargetingNode = GetFirstNodeWithProjectNumber(graph, 2);
+
+            AssertNonMultitargetingNode(nonMultitargetingNode);
+        }
+
+        [Fact]
+        public void NonOuterBuildProjectsInTheMiddle()
+        {
+            var root = CreateProjectFile(_env, 1, new[] {2, 3}, null, null, OuterBuildSpecification).Path;
+            CreateProjectFile(_env, 2, new[] {4});
+            CreateProjectFile(_env, 3, new[] {4});
+            CreateProjectFile(_env, 4, null, null, null, OuterBuildSpecification);
+
+            var graph = new ProjectGraph(root);
+
+            var dot = graph.ToDot();
+
+            graph.ProjectNodes.Count.ShouldBe(8);
+
+            AssertOuterBuildAsRoot(graph.GraphRoots.First());
+            AssertOuterBuildAsNonRoot(GetNodesWithProjectNumber(graph, 4).First(IsOuterBuild));
+
+            AssertNonMultitargetingNode(GetFirstNodeWithProjectNumber(graph, 2));
+            AssertNonMultitargetingNode(GetFirstNodeWithProjectNumber(graph, 3));
+        }
+
+        [Fact]
+        public void InnerBuildsCanHaveSeparateReferences()
+        {
+            var extraInnerBuildReferenceSpec = OuterBuildSpecification +
+                                          $@"<ItemGroup>
+                                                <ProjectReference Condition=`'$({InnerBuildProperty})'=='b'` Include=`4.proj;5.proj`/>
+                                            </ItemGroup>".Cleanup();
+
+            var root = CreateProjectFile(_env, 1, new[] {2, 3}, null, null, extraInnerBuildReferenceSpec).Path;
+            CreateProjectFile(_env, 2, null, null, null, OuterBuildSpecification);
+            CreateProjectFile(_env, 3);
+            CreateProjectFile(_env, 4, null, null, null, OuterBuildSpecification);
```

Comment:
Why does this not produce 4outer->4innerA,4innerB edges?

![image](https://user-images.githubusercontent.com/3347530/54762835-d9b88d00-4bc2-11e9-9704-2034555671bd.png)

Reply:
Answered in https://github.com/Microsoft/msbuild/pull/4218#discussion_r267830872 right?

Reply:
Yes